### PR TITLE
Massive performance boost by disabling Physics Updates during the TMX map import process

### PR DESCRIPTION
**Pull request by elexis1**
**TLDR:** 10 minute vs 15 second performance difference when Physics updates are disabled while importing a TMX file with 2000+ colliders.

We we have been facing devastating import times of more than 10(!) minutes per TMX map file.

This time is spent not only when updating a map but also when changing a TSX file that is used by the map.

Since TSX files can be used by several maps, it can mean that there is more than half an hour to wait for this process to finish, which is obviously inacceptable, especially considering that the expected import time is something around a second.

Trial and error has shown that the importing time is proportional to the number of Colliders placed in the map.
The map that spends 10+ minutes on import has more than 2000 (perfect circle) Ellipse colliders.
Changing the "Tiles as Objects" setting and the number of "Edges Per Ellipse" in the TMX import settings does not affect the performance at all.

We know from Unity Physics that each time a Static Collider is created or moved, the entire Physics simulation state is recomputed - and all colliders imported by Tiled2Unity are Static colliders (because they aren't part of any dynamic or kinematic RigidBody).

Just disabling the Physics simulation entirely during the Asset import fixes the performance issue for us entirely and reduces the import time from 10min+ to 15 seconds!

**Correctness of the patch**:

We know (or can reasonably put that restriction to asset importers) that the asset importing process shall not depend on the Physics simulation state, so temporarily disabling the Physics simulation during the map import should not affect the resulting prefabs.

An alternative idea to addressing this issue might be to add a Dynamic or Kinematic Rigidbody and then remove that later on. But this will most likely be less performant than disabling the Physics simulation entirely.

What seems a bit ugly about this patch:
- If some `SuperImporter` code (e.g. a CustomTmxImporter) would modify `Physics.autoSimulation`, the value would be reset by the last line added in this diff. This is not an issue if we can assume that developers are reasonable enough not to write such code.
- Modifying Physics settings at all during Prefab construction time seems like going out of place of responsibility, since the asset importing code should be independent of the Physics. Evidently it already is heavily dependent on it however.

So if you have a better idea, feel free to ignore this, but we have to patch this locally since we can't use the program while the issue is present.

Avoiding to create that many Colliders is not an option for us currently, since all of these Colliders are meant to be clickable objects, smaller than one tile. Even if authors can avoid creating that many Colliders, it is still a lot of lost performance during map import time for each individual Collider that exists and causes a recomputation of the entire physics simulation state (so even with < 100 colliders it seems worth to patch).

### CODE REVIEW

**Comment by ijacquez** [comment]
Are there calls to import other assets when importing the `.tmx`? If so, [`AssetDatabase.StartAssetEditing`](https://docs.unity3d.com/ScriptReference/AssetDatabase.StartAssetEditing.html) would probably be the "proper" solution.

**Comment by elexis1** [comment]
We've measured the exact amount of time spent per file imported by looking at the `.config/unity3d/Editor.log` logfile during and after the import.

```
Done importing asset: 'Assets/Resources/Maps/WWA_4.1_RECOVERY.tmx' (target hash: '3da0876a985a99b665de45b3a754bc4c') in 821.374753 seconds
```
This individual file with 2000+ Colliders consumes 13 minutes.
If multiple files are involved, the total time consumed of that is a multiple of that.

The current patch doesn't seem to work (no performance difference at all) if "Tiles as Objects" is disabled, I'm trying to locate the reason for that.

**Comment by elexis1** [comment]
**Tesselation:**
The excessive amount of time spent  (several minutes) with `"Tiles as Objects" = false` when there are hundreds or thousands of colliders is due to the Tesselation code and thus unrelated to the Physics simulation and the issue addressed in this patch (`LibTessDotNet/Tess.cs`, `tess.Tessellate` in `TrianglesFromTessellator` and `Extract the triangles` each consumed something like 40% of the total import time each)

**Irreproducible reported issue:**

I have experienced this excessive amount of import time with `"Tiles as Objects" = true` that went away with this patch for several weeks, and spent the past 2 days exclusively measuring it with surgical precision. I've created a minimal reproduction case to upload here. Once I finished that, I could not reproduce it anymore. It is incomprehensible to me how it suddenly disappeared without having changed anything. The only explanation I could come up with is some outdated cache file in Unitys `Library/` folder, but of course can't confirm. I will close the request and reopen it in case I can reproduce it again.

Here my original reply when I could still reproduce the issue.

> **Steps:**
> 1. Create a TSX file with one tile and 1 circular ellipse collider
> 2. Create a TMX file, use bucket fill to place this tile on each part of the grid.
> 3. Measure the exact import time by opening the Unity Editor logfile, search for "import".
>     On unix: `grep import .config/unity3d/Editor.log`
> 4. Measure with "Tiles as Objects" = true and = false. Measure with mapsize 5x5, 7x7, 10x10 and 15x15. Measure with and without the patch.
> 
> Here the minimal TMX and TSX files from Step 1 and Step 2 using only 1 sprite from the SuperTiled2Unity examples (you might have to adapt the filepaths):
> [test.zip](https://github.com/Seanba/SuperTiled2Unity/files/5240164/test.zip)
> 
> **Results (on my machine):**
> 7x7 map without this patch:
> - tiles as objects = false: 7.225965 seconds
> - tiles as objects = true: 3.918272 seconds
> 
> 7x7 map with this patch:
> - tiles as objects = false: 7.046357 seconds
> - tiles as objects = true: 0.917643 seconds
> 
> -----------------------------------------------------
> 
> If you want to see how severe the problem scales with increasing amounts of Colliders, increase the mapsize progressively and suffer:
> 
> With "Tiles as Objects" = true, without this patch:
> - 5x5 map (25 colliders) = 
> - 7x7 map (49 colliders) = 
> - 10x10 map (100 colliders) = 
> - 15x15 map (225 colliders) = 1.5 seconds
> - 50x50 map (2500 colliders) = 3.7 seconds
> - 100x100 map (10k colliders) = 18 seconds
> 
> With "Tiles as Objects" = true and this patch:
> - 5x5 map (25 colliders) = 0.75 seconds
> - 7x7 map (49 colliders) = 0.75 seconds
> - 10x10 map (100 colliders) = 1 second
> - 15x15 map (225 colliders) = 1 second
> - 50x50 map (2500 colliders) = 3.7 seconds 
> 
> With "Tiles as Objects" = false and this patch:
> - 5x5 map (25 colliders) = 1 second
> - 7x7 map (49 colliders) = 3 seconds
> - 10x10 map (100 colliders) = 7 seconds
> - 15x15 map (225 colliders) = 35 seconds
> - 2000+ colliders = beyond 13 minutes
>
